tracepoint_decode/src/perf_field_format.rs

```diff
@@ -0,0 +1,1134 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+extern crate alloc;
+
+use core::ops;
+
+use alloc::string;
+
+use crate::*;
+use eventheader_types::*;
+
+/// At present, does nothing.
+macro_rules! debug_eprintln {
+    ($($arg:tt)*) => {};
+}
+
+/// The type of the array property of [`PerfFieldFormat`].
+/// Array-ness of a field.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum PerfFieldArray {
+    /// e.g. `char val; size:1;`.
+    None,
+
+    /// e.g. `char val[12]; size:12;`.
+    Fixed,
+
+    /// e.g. `char val[]; size:0;`.
+    RestOfEvent,
+
+    /// e.g. `__rel_loc char val[]; size:2;`.
+    /// Value contains relativeOffset. dataLen is determined via strlen.
+    RelLoc2,
+
+    /// e.g. `__data_loc char val[]; size:2;`.
+    /// Value contains offset. dataLen is determined via strlen.
+    DataLoc2,
+
+    /// e.g. `__rel_loc char val[]; size:4;`.
+    /// Value contains `(dataLen << 16) | relativeOffset`.
+    RelLoc4,
+
+    /// e.g. `__data_loc char val[]; size:4;`.
+    /// Value contains `(dataLen << 16) | offset`.
+    DataLoc4,
+}
+
+/// Stores decoding information about a field, parsed from a tracefs "format" file.
+#[derive(Debug)]
+pub struct PerfFieldFormat {
+    field: string::String,
+    name_range: ops::Range<usize>,
+    offset: u16,
+    size: u16,
+    signed: Option<bool>,
+    specified_array_count: u16,
+    deduced_array_count: u16,
+    specified_encoding: FieldEncoding,
+    deduced_encoding: FieldEncoding,
+    specified_format: FieldFormat,
+    deduced_format: FieldFormat,
+    array: PerfFieldArray,
+    element_size_shift: u8,
+}
+
+impl PerfFieldFormat {
+    /// Initializes a `PerfFieldFormat` from a `field` definition plus pre-parsed values for
+    /// other properties. Normally you'll call `parse` to parse a field definition line instead
+    /// of calling this method directly.
+    ///
+    /// Initializes `field`, `offset`, `size`, and `signed` properties exactly as specified.
+    /// Parses `field` to deduce the other properties. The signed parameter should be `None`
+    /// if the "signed:" property is not present in the format line.
+    pub fn new(
+        long_is_64_bits: bool,
+        field: &str,
+        offset: u16,
+        size: u16,
+        signed: Option<bool>,
+    ) -> Self {
+        const SIZEOF_U8: u16 = 1;
```

Comment:
These seem unnecessary?

Reply:
It's a lot easier to read than `mem::size_of::<u8>() as u16`.